File: spring-boot-commons-scripting/src/test/java/com/github/springframework/boot/commons/groovy/GroovyScriptExecutorTest.java
package com.github.springframework.boot.commons.groovy;

import com.github.springframework.boot.commons.groovy.exception.GroovyObjectInvokeMethodException;
import com.github.springframework.boot.commons.groovy.exception.GroovyScriptParseException;
import com.github.springframework.boot.commons.groovy.exception.InvalidGroovyScriptException;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.platform.commons.util.ClassLoaderUtils.getClassLoader;

class GroovyScriptExecutorTest {

	static GroovyScriptExecutor groovyScriptExecutor;

	static String testNormalGroovyScript;

	static String readGroovyScriptAsString(final String fileName) throws IOException {
		ClassLoader classLoader = getClassLoader(GroovyScriptExecutorTest.class);
		try (InputStream inputStream = classLoader.getResourceAsStream(fileName)) {
			assert inputStream != null;
			InputStreamReader streamReader = new InputStreamReader(inputStream);
			BufferedReader bufferedReader = new BufferedReader(streamReader);
			return bufferedReader.lines().collect(Collectors.joining("\n"));
		}
	}

	@BeforeAll
	static void setup() throws IOException {
		groovyScriptExecutor = new GroovyScriptExecutor();
		testNormalGroovyScript = readGroovyScriptAsString("TestGroovyScriptExecutor.groovy");
	}

	@Test
	void testGroovyScriptNullOrEmpty() {
		// null
		InvalidGroovyScriptException nullException = assertThrows(InvalidGroovyScriptException.class,
			() -> groovyScriptExecutor.execute(null, "test", 1)
		);
		assertEquals("Groovy script is null", nullException.getMessage());
		// empty
		InvalidGroovyScriptException emptyException = assertThrows(InvalidGroovyScriptException.class,
			() -> groovyScriptExecutor.execute("", "test", 1)
		);
		assertEquals("Groovy script is empty", emptyException.getMessage());
	}

	@Test
	void testGroovyScriptInvokeMethodNoArgs() {
		Object result = groovyScriptExecutor.execute(testNormalGroovyScript, "testInvokeMethodNoArgs");
		assertEquals(1024, result);
	}

	@Test
	void testGroovyScriptInvokeMethodWithArgs() {
		Object result = groovyScriptExecutor.execute(testNormalGroovyScript, "testInvokeMethodWithArgs", 1);
		assertEquals(2, result);
	}

	@Test
	void testGroovyScriptInvokeMethodWithTwoArgs() {
		Object result = groovyScriptExecutor.execute(testNormalGroovyScript, "testInvokeMethodWithTwoArgs", 1, 2);
		assertEquals(3, result);
	}

	@Test
	void testGroovyScriptInstantiationException() throws IOException {
		final String script = readGroovyScriptAsString("TestInstantiationException.groovy");
		GroovyScriptParseException e = assertThrows(GroovyScriptParseException.class,
			() -> groovyScriptExecutor.execute(script, "test")
		);
		assertTrue(e.getMessage().contains("Failed to parse groovy class script"));
	}

	@Test
	void testEvaluateGroovyScriptSnippet() {
		final String scriptSnippet = "def test() { return 1 + 1 }";
		Object result = groovyScriptExecutor.evaluate(scriptSnippet, "test");
		assertEquals(2, result);
		// case: groovyObjectCache != null
		result = groovyScriptExecutor.evaluate(scriptSnippet, "test");
		assertEquals(2, result);
	}

	@Test
	void testGroovyObjectInvokeMethodException() {
		final String script = "def testAdd() { return 1 + 1 }";
		GroovyObjectInvokeMethodException e = assertThrows(GroovyObjectInvokeMethodException.class,
			() -> groovyScriptExecutor.evaluate(script, "test")
		);
		assertTrue(e.getMessage().contains("Failed to invoke groovy method"));
	}

	@Test
	void testGroovySecurityInterceptor() {
		// unsafe script case
		final String unsafeScript1 = "def testSystemClassInterceptor() { System.gc() }";
		GroovyObjectInvokeMethodException e = assertThrows(GroovyObjectInvokeMethodException.class,
			() -> groovyScriptExecutor.evaluate(unsafeScript1, "testSystemClassInterceptor")
		);
		assertTrue(e.getMessage().contains("not allowed in your groovy code"));

		// another unsafe script case
		final String unsafeScript2 = "def testRuntimeClassInterceptor() { Runtime.getRuntime().gc() }";
		e = assertThrows(GroovyObjectInvokeMethodException.class,
			() -> groovyScriptExecutor.evaluate(unsafeScript2, "testRuntimeClassInterceptor")
		);
		assertTrue(e.getMessage().contains("not allowed in your groovy code"));

		// safe script case
		final String safeScript = "def testSystemClassInterceptor() { return System.currentTimeMillis() }";
		Object result = groovyScriptExecutor.evaluate(safeScript, "testSystemClassInterceptor");
		assertTrue(result instanceof Long && Long.parseLong(result.toString()) > 0);
	}

}

File: spring-boot-commons-scripting/src/test/java/com/github/awesome/springboot/commons/scripting/groovy/GroovyScriptExecutorTest.kt
package com.github.awesome.springboot.commons.scripting.groovy

import com.github.awesome.springboot.commons.scripting.groovy.exception.GroovyObjectInvokeMethodException
import com.github.awesome.springboot.commons.scripting.groovy.exception.GroovyScriptParseException
import com.github.awesome.springboot.commons.scripting.groovy.exception.InvalidGroovyScriptException
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class GroovyScriptExecutorTest {

    private val groovyScriptExecutor = GroovyScriptExecutor()

    @Test
    fun testExecute() {
        val e1 = assertThrows<InvalidGroovyScriptException>() {
            groovyScriptExecutor.execute(null, null)
        }
        assertEquals("Groovy script is null", e1.message)

        val e2 = assertThrows<InvalidGroovyScriptException>() {
            groovyScriptExecutor.execute("", null)
        }
        assertEquals("Groovy script is empty", e2.message)

        val groovyClassScript = """
            class GroovyClass {
                def execute() {
                    return "Hello, Groovy!"
                }

                def executeWithParam(String name) {
                    return "Hello, " + name
                }
            }
        """
        val result1 = groovyScriptExecutor.execute(groovyClassScript, "execute")
        assertEquals("Hello, Groovy!", result1)

        val result2 = groovyScriptExecutor.execute(groovyClassScript, "executeWithParam", "Groovy!")
        assertEquals("Hello, Groovy!", result2)

        val e3 = assertThrows<GroovyScriptParseException>() {
            groovyScriptExecutor.execute("interface GroovyInterface {}", "execute")
        }
        assertTrue { e3.message!!.startsWith("Failed to parse groovy class script") }

        val e4 = assertThrows<GroovyObjectInvokeMethodException>() {
            groovyScriptExecutor.execute(groovyClassScript, "noSuchMethod")
        }
        assertTrue { e4.message!!.startsWith("Failed to invoke groovy method") }
    }

    @Test
    fun testEvaluate() {
        val e1 = assertThrows<InvalidGroovyScriptException>() {
            groovyScriptExecutor.evaluate(null, null)
        }
        assertEquals("Groovy script is null", e1.message)

        val e2 = assertThrows<InvalidGroovyScriptException>() {
            groovyScriptExecutor.evaluate("", null)
        }
        assertEquals("Groovy script is empty", e2.message)

        val groovyScript = """
            def execute() {
                return "Hello, Groovy!"
            }

            def executeWithParam(String name) {
                return "Hello, " + name
            }
        """
        val result1 = groovyScriptExecutor.evaluate(groovyScript, "execute")
        assertEquals("Hello, Groovy!", result1)

        val result2 = groovyScriptExecutor.evaluate(groovyScript, "executeWithParam", "Groovy!")
        assertEquals("Hello, Groovy!", result2)

        val e3 = assertThrows<GroovyObjectInvokeMethodException>() {
            groovyScriptExecutor.evaluate(groovyScript, "noSuchMethod")
        }
        assertTrue { e3.message!!.startsWith("Failed to invoke groovy method") }
    }

    @Test
    fun testGroovyInterceptor() {
        val groovyScript = """
            def unsafeSystemMethod() {
                System.gc()
            }

            def safeSystemMethod() {
                System.currentTimeMillis()
            }

            def unsafeRuntimeClass() {
                Runtime.getRuntime().exec("ls")
            }
        """
        val e1 = assertThrows<GroovyObjectInvokeMethodException>() {
            groovyScriptExecutor.evaluate(groovyScript, "unsafeSystemMethod")
        }
        assertTrue { e1.message!!.contains("not allowed in your groovy code") }

        val result1 = groovyScriptExecutor.evaluate(groovyScript, "safeSystemMethod")
        assertTrue { result1 is Long && result1 > 0 }

        val e2 = assertThrows<GroovyObjectInvokeMethodException>() {
            groovyScriptExecutor.evaluate(groovyScript, "unsafeRuntimeClass")
        }
        assertTrue { e2.message!!.contains("not allowed in your groovy code") }
    }

}
